Add getting started with xrpl-py tutorial

ryangyoung opened the pull request:
Complete draft of a basic getting started tutorial, including sample code. 

See a preview of this on staging: https://developers-staging.ripple.com/get-started-using-python.html

ryangyoung committed:
initial structure and first code samples

ryangyoung committed:
add code samples for connect and query steps

ryangyoung committed:
add new independent py tutorial and env setup

ryangyoung committed:
add v rough sample code and initial structure

mvadari commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,97 @@
+        seed=_SEED,
+        seed_hex=_SEED_HEX,
+    )
+    print("Signed transaction: ", signed_tx)
```
This is a model that is used to send a `Sign` request to the API, not one that actually signs the object (for that you'll need the keypairs method `sign` and also to encode it via the binary codec). See https://github.com/xpring-eng/xrpl-py/blob/master/xrpl/transaction/main.py#L34

mvadari replied:
Looks like this part isn't actually being used anywhere.

mvadari commented on content/tutorials/get-started/get-started-build-apps.md:
````diff
@@ -0,0 +1,177 @@
+*Python (xrpl-py)*
+
+```py
+account_info = RequestMethod.ACCOUNT_INFO(rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe)
````
```suggestion
account_info = RequestMethod.ACCOUNT_INFO(account="rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe")
```
All the models require keyword args

mvadari commented on content/tutorials/get-started/get-started-build-apps.md:
```diff
@@ -0,0 +1,177 @@
+
+## Submit transaction
+
+When you're ready 
```
See https://github.com/xpring-eng/xrpl-py/blob/master/xrpl/transaction/main.py#L56

mvadari commented on content/tutorials/get-started/get-started-with-python-sdk.md:
````diff
@@ -0,0 +1,202 @@
+```py
+# Generate private and public keys
+# to manage your XRP Ledger account
+def generateKeys():
````
```suggestion
def generate_keys():
```
Python typically uses snake_case, not CamelCase

mvadari replied:
This should probs be true for all your method names.

mvadari commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,97 @@
+    tx = Transaction(
+        account=_ACCOUNT,
+        fee=_FEE,
+        sequence=CLASSIC_ACCOUNT.next_seuqnece_num + 10,
```
This is supposed to be just `next_sequence_num` - it's the next sequence number, not the last ledger sequence value.

mDuo13 replied:
Also note the typo (`seuqnece`)

mvadari commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,97 @@
+        account=_ACCOUNT,
+        fee=_FEE,
+        sequence=CLASSIC_ACCOUNT.next_seuqnece_num + 10,
+        transaction_type=TransactionType.
```
I assume this is supposed to be a type of transaction?

mvadari commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,97 @@
+        sequence=CLASSIC_ACCOUNT.next_seuqnece_num + 10,
+        transaction_type=TransactionType.
+    )
+    value = tx.to_dict()["payment_trannsaction"]
```
Not really sure what this is supposed to be doing, there isn't any `"payment_transaction"` value that will be generated in the dictionary.

mvadari commented on content/tutorials/get-started/get-started-with-python-sdk.md:
```diff
@@ -0,0 +1,202 @@
+        payment_transaction, CLASSIC_ACCOUNT, JSON_RPC_URL
+        )
+    print("response.result.validated: ", response["validated"])
+    print("response.result.meta: ", response.result["meta"]["TransactionResult"], "tesSUCCESS")
```
See comments above

mvadari commented on content/tutorials/get-started/get-started-with-python-sdk.md:
```diff
@@ -0,0 +1,202 @@
+if response.result["meta"]["TransactionResult"] != "tesSUCCESS":
+    print("Transaction not yet validated")
+else:
+    print("Validated!")  
```
There is also `response.is_successful()`

mDuo13 replied:
Doesn't `response.is_successful()` check whether the server received and understood the question? That's a completely different thing than whether the transaction executed with a result code of `tesSUCCESS`.

mvadari commented on content/tutorials/get-started/get-started-with-python-sdk.md:
```diff
@@ -0,0 +1,202 @@
+
+### {{n.next()}}. Verify results
+
+To ensure that your transaction has succeeded in a validated ledger, which means that the results are final, you can add simple logic to your submit function:
```
You may want to point out the difference between using a standard `submit` vs the `send_reliable_submission` above - without that specific method that does the heavy lifting, it doesn't double check validation.

mDuo13 commented on dactyl-config.yml:
```diff
@@ -1579,6 +1579,14 @@ pages:
         targets:
             - ja
 
+    -   md: tutorials/get-started/get-started-with-python-sdk.md
```
As much as possible, I try to keep the following things consistent with each other:

- The actual title displayed on the page
- The MD filename the content comes from
- The HTML filename the content goes to

So when you're reasonably confident you have a final title, you should try to make the three match. So for example:

- Get Started Using Python
- `get-started-using-python.md`
- `get-started-using-python.html`

My convention is to include all the words of the title so that I don't have to think about which words are unimportant enough to omit. (That's subjective, so remembering how to do it consistently is hard.)

mDuo13 commented on content/tutorials/get-started/get-started-with-python-sdk.md:
```diff
@@ -0,0 +1,202 @@
+cta_text: Build Apps
+---
+
+# Get started using Python
```
By convention, page titles use title case.

```suggestion
# Get Started Using Python
```

mDuo13 commented on content/tutorials/get-started/get-started-with-python-sdk.md:
```diff
@@ -0,0 +1,202 @@
+
+In this tutorial, you'll learn: 
+
+* How to set up your environment for Python development. See [](xref:python-env-setup.md).
```
We're not using XREFs in the dev portal right now, so this shows up as a blank or broken link, as do the other `xref` examples. (Whether or not we _should_ be using xrefs is a valid question, but for now you should probably just make this link manual.)

mDuo13 commented on content/tutorials/get-started/get-started-with-python-sdk.md:
```diff
@@ -0,0 +1,202 @@
+4. [Submit a transaction.](#submit-transaction) 
+5. [Verify results.](#verify-results) 
+
+### {{n.next()}}.  Generate keys
```
Our "Quickstart" guide should jump straight to querying the XRP Ledger as soon as possible. Since key generation is only necessary for sending transactions, I think we should move this aside and put it in a "Sending your first transaction" tutorial or sub-tutorial.

mDuo13 commented on content/tutorials/get-started/get-started-with-python-sdk.md:
```diff
@@ -0,0 +1,202 @@
+# Generate private and public keys
+# to manage your XRP Ledger account
+def generateKeys():
+    seed = keypairs.generate_seed()
```
Where's the line to import the module? The very beginning of any code snippet in our Quick Start should be something like:

```
import xrpl
```

This is particularly important because the names of the modules you import are often not the same as the name of the package you installed, for various reasons. (For example, you can't have `-` in variable names in Python, since that's the minus operator, so `xrpl-py` would not be a valid name for a thing you could import.)

This line, for example,  isn't going to work without some previous code to define `keypairs`. Is that imported from the `xrpl` library somehow? Or instantiated/returned from some class in the library?

mDuo13 commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,97 @@
+CLASSIC_ACCOUNT = ""
+SENDER_ACCOUNT = "rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe"
+address = ""
+DESTINATION_WALLET = generate_faucet_wallet(JSON_RPC_URL).classic_address
```
`generate_faucet_wallet` doesn't exist yet here. You're missing an import:

```
from xrpl.wallet import generate_faucet_wallet
```

ledhed2222 commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,97 @@
+import xrpl
```
no need to import this :)

ryangyoung committed:
updated sample app per teting and feedback

mvadari commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,151 @@
+_SECRET = "randomsecretkey"
+_SEED = "randomsecretseedforakey"
+_SEED_HEX = "EACEB081770D8ADE216C85445DD6FB002C6B5A2930F2DECE006DA18150CB18F6"
+_PASSPHRASE = "mytopsecretpassphrasethatwillneverbehacked"
```
Should these be the seeds that are generated below?

mvadari commented on content/_code-samples/xrpl-py/simple-python-app.py:
```diff
@@ -0,0 +1,151 @@
+    payment_transaction = Payment.from_dict(value)
+    response = send_reliable_submission(
+        payment_transaction, CLASSIC_ACCOUNT, JSON_RPC_URL
+        )
```
This is going to resend the transaction. We don't currently have a function that checks an already submitted transaction (though it wouldn't be hard to quickly write if you think that's necessary).

mvadari commented on content/tutorials/get-started/get-started-build-apps.md:
````diff
@@ -0,0 +1,177 @@
+
+```py
+seed = keypairs.generate_seed()
+public, private = keypairs.derive_keypair(seed)
````
Should this include the classic address derivation?

mvadari commented on content/tutorials/get-started/get-started-build-apps.md:
````diff
@@ -0,0 +1,177 @@
+*Python (xrpl-py)*
+
+```py
+client = JsonRpcClient(http://s1.ripple.com:51234/)
````
```suggestion
client = JsonRpcClient("http://s1.ripple.com:51234/")
```

mDuo13 commented on content/tutorials/get-started/python-env-setup.md:
```diff
@@ -0,0 +1,116 @@
+
+## Python Environment Setup
+
+Complete the steps in the following sections to prepare your environment for development. 
```
The steps described in this page are all helpful for developing/contributing to the xrpl-py library itself, but are not at all required for merely _using_ the library. In my opinion, instructions for developing the library belong in the library's repo and it's not necessary to duplicate them on xrpl.org.

I'm reasonably confident that the typical install will be as simple as running `pip3 install xrpl-py` (as long as you have Python 3.7 or newer) so we won't even need a full separate page for the installation/setup steps, and we can remove this entire page from here. It may be useful to contribute these changes back to the repo itself as a contributing guide though.

mDuo13 reviewed (changes requested):
Just realized I forgot to submit my feedback from earlier. Apologies for any comments that are out of date.

ryangyoung committed:
revise sample app and tutorial

ryangyoung committed:
remove env setup, begin refactor

ryangyoung committed:
finish revision, use code includes

ryangyoung committed:
rebase and add include_code to page

ryangyoung commented:
~When I try to build this locally, I get the following error:~

```py
Building page: <Page 'Get Started Using Python' (get-started-using-python.html)>
  File "/usr/local/lib/python3.9/site-packages/dactyl/page.py", line 215, in preprocess
    md = self.pp_template.render(**context, **self.filter_exports())
  File "/usr/local/lib/python3.9/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/usr/local/lib/python3.9/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/local/lib/python3.9/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "content/tutorials/get-started/get-started-using-python.md", line 158, in top-level template code
  File "/usr/local/lib/python3.9/site-packages/dactyl/jinja_loaders.py", line 54, in get_source
    text, filename, uptodate = super().get_source(environment, template)
  File "/usr/local/lib/python3.9/site-packages/jinja2/loaders.py", line 197, in get_source
    raise TemplateNotFound(template)
Preprocessor error in page <Page 'Get Started Using Python' (get-started-using-python.html)>: _code-samples/xrpl-py/get-acct_info.py.
```

~Seems like Dactyl is complaining about a missing template for the `get-acct_info.py` and I'm not sure why. I rebased to get the `include_code` filter; not sure if this error is at all related. @mDuo13 any ideas on what might going on?~


The error was due to a few syntax errors and typos:

* `include_code` does not require code fences. Removed those from the included examples. 
* The file name for the complete sample had an `_` instead of a `-`
* Several `include_code` samples had an extra `)` at the end

These have all been corrected and the site builds correctly now.

ryangyoung commented:
@mvadari @mDuo13 @ledhed2222 I completed a major overhaul for this tutorial:

* Radically simplified the app to only generate a wallet and look up account info for that wallet
* Revised the text of the tutorial to match the scope of the app
* Tried to use `include_code` to pull in code samples from the app (this seems to be broken now)
* Added a link to the Send XRP tutorial, which Rome has [updated with Python examples](https://github.com/ripple/xrpl-dev-portal/pull/1009)

This revision should resolve all of the existing comments or make outdated (I've tried to reflect this in the diff). 

Next, I'll try to update the secure signing page with Python examples, and will look for other spots where I might  be able to do the same. And revise based on all of your feedback, of course :)

ryangyoung committed:
fix include_code syntax errors

mvadari commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+seed: shrPSF6vV3v3yoND9J3NeKks6M3xd
+pub_key: 022FA613294CD13FFEA759D0185007DBE763331910509EF8F1635B4F84FA08AEE3
+priv_key: 00D46D49A1A260C24964E6CFD5C85308ACE958E200C14B4DCA73AA5C87288C8F6C
+classic_address: raaFKKmgf6CRZttTVABeTcsqzRQ51bNR6Q
```
This output will be changed in https://github.com/XRPLF/xrpl-py/pull/194

ryangyoung replied:
Updated this to be:

```
# print output
public_key:: 022FA613294CD13FFEA759D0185007DBE7633319105
private_key:: -HIDDEN-
classic_address: raaFKKmgf6CRZttTVABeTcsqzRQ51bNR6Q
```

mvadari commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+
+#### Using the wallet
+
+In this tutorial we only query details about the generated account from the XRP Ledger, but you can use the values in the `Wallet` instance to prepare, sign, and submit transactions with `xrpl-py`.   
```
This process will likely be modified in a couple of future PRs (including https://github.com/XRPLF/xrpl-py/pull/197)

ryangyoung replied:
Looks like https://github.com/XRPLF/xrpl-py/pull/197 isn't merged yet, so I'll leave this as-is for now and will update this after the merge happens.

ryangyoung replied:
Updated these per that PR and a few other changes.

mvadari reviewed (commented):
This looks awesome! 

Heads up, there's going to be a few public-facing API changes in the next couple of days that will affect the actual code here.

mDuo13 commented on dactyl-config.yml:
```diff
@@ -1580,6 +1580,10 @@ pages:
+    -   md: tutorials/get-started/get-started-using-python.md
+        targets:
+            - en
+
```
Add new pages to the `- ja` target too until they're translated. (Otherwise you get a broken links to the Japanese version of the page)

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+
+1. [Connect to the XRP Ledger.](#connect)
+1. [Generate a wallet.](#generate-wallet)
+1. [Query the XRP Ledger.](#query)
```
These links don't work because the numbering is part of the anchors. So, as they are now, they should be:

```suggestion
1. [Connect to the XRP Ledger.](#1-connect)
1. [Generate a wallet.](#2-generate-wallet)
1. [Query the XRP Ledger.](#3-query)
```

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+
+You can query the XRP Ledger to get information about [a specific account](account-methods.html), [a specific transaction](tx.html), the state of a [current or a historical ledger](ledger-methods.html), and [the XRP Ledger's decentralized exhange](path-and-order-book-methods.html). You need to make these queries, among other reasons, to look up account info to follow best practices for [reliable  ...
+
+Here, we'll use `xrpl-py`'s [`xrpl.account`](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.account.html) module to look up information about the [wallet we generated](#generate-wallet) in the previous step. 
```
The `#generate-wallet` link is broken here too, for the same reason as above.

```suggestion
Here, we'll use `xrpl-py`'s [`xrpl.account`](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.account.html) module to look up information about the [wallet we generated](#2-generate-wallet) in the previous step. 
```

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+
+### Requirements for contributing to the library
+
+If you want to contribute code to the library itself, see [CONTRIBUTING](https://github.com/XRPLF/xrpl-py/blob/master/CONTRIBUTING.md) in the project repo. 
```
I don't think this belongs here. I'd worry it's more likely to side-track people than help them. When someone is just getting started using a library, contributing back to the development on that library is usually pretty far away.

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+
+### {{n.next()}}. Connect
+
+To make queries and submit transactions, you need to establish a connection to the XRP Ledger. To do this with `xrpl-py`, use the [`xrp.clients` module](https://xrpl-py.readthedocs.io/en/latest/source/xrpl.clients.html):
```
It would be good to add a note before or after the code sample clarifying that this example uses a public Testnet server URL. (In the Send XRP tutorial, there's a mention of this at the equivalent point, and also [an entire section on how to connect to Mainnet](https://xrpl.org/send-xrp.html#connecting-to-the-production-xrp-ledger).)

ryangyoung replied:
Added a section to match the example in the other tutorial. This might be a good candidate for its own page/min-tutorial in the future.

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+
+my_tx_payment = Payment(
+    account=test_wallet.classic_address,
+    amount="2200", # in drops: https://xrpl.org/basic-data-types.html#specifying-currency-amounts
```
The reason the Send XRP tutorial uses `"22000000"` (a whole 22 XRP) is that it's slightly more than enough to fund an account even if the user supplies a different destination address.

You could also use `xrpl.utils.xrp_to_drops()` here now that that's merged.

mDuo13 commented on content/_code-samples/xrpl-py/get-acct-info.py:
```diff
@@ -0,0 +1,35 @@
+# Derive an x-address from the classic address:
+# https://xrpaddress.info/
+from xrpl.core import addresscodec
+test_xaddress = addresscodec.classic_address_to_xaddress(test_account, tag=0, is_test_network=True)
```
This example doesn't make it clear that tag=0 is different than not having a tag. Maybe use something that's more obviously an example tag, like `12345`, instead of `0`.

See the following comparison of X-addresses for the faucet's classic address rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe:

| Test? | Tag | Resulting X-Address |
|---|---|---|
| Yes | None | TVeG6DZkiwxmsLrXeYEu5WuSxJDWudgMDR839hr9vgTeK2C |
| Yes | 0 | TVeG6DZkiwxmsLrXeYEu5WuSxJDWudmBQ5NjFeRakePCJ71 |
| No | None | XV5kHfQmzDQjbFNv4jX3FX9Y7ig5QhpKGEFCq4mdLfhdxMq |
| No | 0 | XV5kHfQmzDQjbFNv4jX3FX9Y7ig5Qhf9TYHHWyB2wsNKc6Y |

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
````diff
@@ -0,0 +1,215 @@
+```sh
+git clone git@github.com:ripple/xrpl-dev-portal.git
+cd path/to/xrpl-dev-portal/content/_code-samples/xrpl-py/get-acct-info.py
+python3 get-acct-info.py 
````
Rather than cloning the entire repository to get one file, this is an example where the widely-available `wget` tool would be faster and less tricky:

```sh
$ wget https://raw.githubusercontent.com/ripple/xrpl-dev-portal/xrpl-py-tutorial/content/_code-samples/xrpl-py/get-acct-info.py
$ python3 get-acct-info.py
```

Though maybe it doesn't set the greatest example to ask people to download a random script off the internet and then immediately run it?

ryangyoung replied:
Hmm, `wget` is not installed by default on macos. Easy enough to install via `brew` but that seems like a lot of steps to get one file. 

You could also get the content via `curl`:

```
curl -0  "https://raw.githubusercontent.com/ripple/xrpl-dev-portal/xrpl-py-tutorial/content/_code-samples/xrpl-py/get-acct-info.py"
```

But that only writes it to `stdout`. I think you could also tweak that to write to a specific file locally but I haven't been able to get that to work.

mDuo13 replied:
`git` isn't installed by default on macOS, either, though you're right that git is more likely to be useful for other things anyway

mDuo13 replied:
Since the xrpl-dev-portal repo will get cloned into a folder of the same name relative to where you are when you run the `git clone` command, the following command should consistently work:

```suggestion
git clone git@github.com:ripple/xrpl-dev-portal.git
cd xrpl-dev-portal/content/_code-samples/xrpl-py/get-acct-info.py
python3 get-acct-info.py
```

mDuo13 commented on content/_code-samples/xrpl-py/get-acct-info.py:
```diff
@@ -0,0 +1,35 @@
+from xrpl.models.requests.account_info import AccountInfo
+acct_info = AccountInfo(
+    account=test_account,
+    ledger_index="current",
```
I suggest using `"validated"` here

phoenix7365 replied:
Nah bro it sucks

mDuo13 commented on content/_code-samples/xrpl-py/get-acct-info.py:
```diff
@@ -0,0 +1,35 @@
+# https://xrpaddress.info/
+from xrpl.core import addresscodec
+test_xaddress = addresscodec.classic_address_to_xaddress(test_account, tag=0, is_test_network=True)
+print("\nClassic address:\n\n", test_account)
+print("X-address:\n\n", test_xaddress)
```
Just a stylistic thing, but I don't think it's all that helpful to add explicit linebreaks to these print statements.

ryangyoung replied:
Fair. I'm gonna leave them in for now just because it does make the output a little easier to read, though it's not super complicated, of course.

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
````diff
@@ -0,0 +1,215 @@
+python3 get-acct-info.py 
+```
+
+You should see output similar to this example:
````
One place to expand on this would be to talk a little about what these results mean.

- Call out a few salient points here, like the sequence number (`account_data.Sequence`), XRP balance in drops (`account_data.Balance`), and whether the details are/aren't from a validated ledger.
    - This could be a good place to show off `xrpl.utils.drops_to_xrp()`
- Link to [account_info](https://xrpl.org/account_info.html) for the full definition of the response format.

mDuo13 commented on content/tutorials/get-started/get-started-using-python.md:
```diff
@@ -0,0 +1,215 @@
+
+* [Send XRP](send-xrp.html).
+* [Set up secure signing](set-up-secure-signing.html) for your account.
+* Set an [account flag](accountset.html#accountset-flags).
```
I'd leave this off for now since it doesn't provide a clear path to action. But I've opened #1013 to cover an article that does provide more direction on that front.

mDuo13 reviewed (changes requested):
Looks awesome! I suggested a couple changes to fix broken links and simplify a couple other parts, but overall this looks very good already (and I can confirm that the code works).

phoenix7365 commented:
Great! Where is my Free XRP!?

phoenix7365 commented:
I'm a very valuable African Prince XRP investor, I demand you to send me 1,000 real XRP to test the code sequence of all the above!
Thanks,
Prince of Kenya

ryangyoung committed:
revise per Rome, Mayuhka feedback

ryangyoung committed:
add validated explanation

ryangyoung committed:
change current to validated for ledger

ryangyoung commented on content/_code-samples/xrpl-py/get-acct-info.py:
```diff
@@ -0,0 +1,35 @@
+from xrpl.models.requests.account_info import AccountInfo
+acct_info = AccountInfo(
+    account=test_account,
+    ledger_index="validated",
```
@mDuo13 I changed this to validated but testing code elsewhere, I got an error when trying to use `"validated"`:

```py
>>> acct_info = AccountInfo(
...   account="rBtXmAdEYcno9LWRnAGfT9qBxCeDvuVRZo",
...     ledger_index="validated",
...       queue=True,
...     strict=True,
... )
>>> response = client.request(acct_info)
>>> result = response.result
>>> import json
>>> print(json.dumps(result["account_data"], indent=4, sort_keys=True))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'account_data'
>>> print(result)
{'error': 'invalidParams', 'error_code': 31, 'error_message': 'Invalid parameters.', 'ledger_hash': '78D93327787EFF11BD5AB6507BF81CE5514B422321C798326F954A19E9C33DF6', 'ledger_index': 16234298, 'request': {'account': 'rBtXmAdEYcno9LWRnAGfT9qBxCeDvuVRZo', 'command': 'account_info', 'ledger_index': 'validated', 'queue': True, 'signer_lists': False, 'strict': True}, 'validated': True}
>>> print(response)
Response(status=<ResponseStatus.ERROR: 'error'>, result={'error': 'invalidParams', 'error_code': 31, 'error_message': 'ledgerIndexMalformed', 'request': {'account': 'rBtXmAdEYcno9LWRnAGfT9qBxCeDvuVRZo', 'command': 'account_info', 'ledger_index': 'Validated', 'queue': True, 'signer_lists': False, 'strict': True}}, id=None, type=<ResponseType.RESPONSE: 'response'>)
```

mDuo13 committed:
xrpl-py-tutorial: Fix account_info invalid params, add faucet debug

ryangyoung committed:
remove escrow link

ryangyoung committed:
updatee prepare payment section

ryangyoung committed:
simplify git clone and cd path

mDuo13 committed:
xrpl-py tutorial: fix links in JA target

mDuo13 commented:
Local build reported:

```text
Link check report. 128829 links checked.
Success! No broken links found.
```

So I think we're set here.
